src/status.h

```diff
@@ -0,0 +1,141 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NINJA_STATUS_H_
+#define NINJA_STATUS_H_
+
+#include <map>
+#include <string>
+using namespace std;
+
+#include "build.h"
+#include "line_printer.h"
+
+
+/// Abstract interface to object that tracks the status of a build:
+/// completion fraction, printing updates.
+struct Status {
+  virtual void PlanHasTotalEdges(int total) = 0;
+  virtual void BuildEdgeStarted(Edge* edge) = 0;
+  virtual void BuildEdgeFinished(Edge* edge, bool success, const string& output,
+                         int* start_time, int* end_time) = 0;
+  virtual void BuildLoadDyndeps() = 0;
+  virtual void BuildStarted() = 0;
+  virtual void BuildFinished() = 0;
+
```

Comment:
Unless you plan to somehow prevent deleting a `Status*`, you need to have a virtual destructor. Otherwise, the destructor for derived classes won't be called.

Reply:
If you do plan to prevent deleting a `Status*`, and only allow deleting derived classes, then you should define a non-virtual destructor, and make it `protected` so that Status cannot be directly deleted.

Reply:
No need for protected, deleting Status* without a virtual dtor will result in a warning.

Reply:
You cannot rely on a compiler to issue a diagnostic for this. Not only will some compilers simply not issue a diagnostic for this at all, but even compilers that do may not be able to issue a diagnostic for this in every situation due to AST transformations losing the context of the delete operation.

Using the protected keyword guarantees that the code will not compile if Status* is deleted directly.

Rely on official ISO C++ standard behavior. Not implementation whims of the compiler vendor.

Reply:
> but even compilers that do may not be able to issue a diagnostic for this in every situation due to AST transformations losing the context of the delete operation.

Can you give an example?

Reply:
Of *this specific* situation? No.

But I've seen first hand several times where GCC, or MSVC, would properly warn about a problem with a particular piece of code, and then a change to another function that happened to call the function being warned about resulted in the warning going away.

The only conclusion that I was able to come to when I investigated further is that the compilers decision to inline some aspect of the functions in question resulted in the diagnostic not triggering due to the specific implementation of the diagnostic not being aware of the inlining behavior that was happening.

Regardless, my point is that diagnostics are not standard. Relying on them (e.g. with -Werror) to ensure a particular usage of a part of the code is not allowed is not what the diagnostics are intended for, but there is an explicit, standard defined, mechanism that can be used to ensure this interface class cannot be mis-used.

Ninja should prefer the explicitly standard defined mechanisms over trusting that the compiler vendor implemented a diagnostic, or implemented it for all situations.

Reply:
I disagree.

Reply:
You... Disagree about what, specifically?

Reply:
> Ninja should prefer the explicitly standard defined mechanisms over trusting that the compiler vendor implemented a diagnostic, or implemented it for all situations.

At least in this particular case.

Reply:
This is where the comment that ended up in the other section was intended to go.

> Justify your position please.

> You are ignoring the standard mechanism to prevent bad behavior, which costs 2 lines of code, in favor of relying on arbitrary functionality offered by some compilers, some times, with no guarantee of working, and no guarantee that it'll continue working in future releases.

Reply:
> some times,

How do you know that?

Reply:
Its not a diagnostic that is mandated by the standard. The burden of proof that this diagnostic will be available is on you, not me.

Reply:
See GCC's or Clang's compiler manual.

Reply:
So?

What about MSVC? ICC? Other compilers that I don't know about.

With just two lines of code. We can solve this problem in a manner that is guaranteed by the standard.

Why are you against that?

Reply:
I never said that I'm "against" it, just that there is no need.